In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [6]:
DATA_PATH = "../docs/master_property_tax_enforcement_dataset.csv"

master_property = pd.read_csv(DATA_PATH, low_memory=False)

print(master_property.shape)
display(master_property.head())

(9763, 73)


,month,cycle,borough_lien,block_lien,lot_lien,tax_class_code,building_class,community_board,council_district,house_number,...,vacancy_rate,rent_burden_rate,severe_rent_burden_rate,black_rate,hispanic_rate,asian_rate,white_nonhispanic_rate,age_65_plus_rate,bachelors_plus_rate,matched_acs
0,2019-10-01T00:00:00.000,Final Sale,1,26,1260,2,R4,101.0,1.0,15,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0
1,2019-10-01T00:00:00.000,Final Sale,1,26,1273,2,R4,101.0,1.0,15,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0
2,2019-10-01T00:00:00.000,Final Sale,1,27,1040,2,R4,101.0,1.0,55,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0
3,2019-10-01T00:00:00.000,Final Sale,1,31,1068,2,R4,101.0,1.0,75,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0
4,2019-10-01T00:00:00.000,Final Sale,1,44,1265,2,R4,101.0,1.0,18,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0


#### Section 1: Property Composition

What type of properties dominate the lien sale pool?

In [20]:
# TAX CLASS

tax_class_summary = (
    master_property["tax_class_code"]
    .value_counts(dropna=False)
    .rename_axis("tax_class")
    .reset_index(name="properties")
)

tax_class_summary["share"] = (
    tax_class_summary["properties"] /
    tax_class_summary["properties"].sum()
)

display(tax_class_summary)

,tax_class,properties,share
0,1,4744,0.485916
1,4,2718,0.278398
2,2,2301,0.235686


<B>
The final-sale pool is dominated by residential property. </B>

 Final-sale pool is:

48.6% Class 1 <BR>
23.6% Class 2 <br>
27.8% Class 4 <br>


In [21]:

# Borough composition

borough_summary = (
    master_property["borough_lien"]
    .value_counts(dropna=False)
    .rename_axis("borough")
    .reset_index(name="properties")
)

borough_summary["share"] = (
    borough_summary["properties"] /
    borough_summary["properties"].sum()
)

display(borough_summary)

,borough,properties,share
0,3,3715,0.380518
1,4,2845,0.291406
2,2,1324,0.135614
3,1,1067,0.109290
4,5,812,0.083171


<b>
Brooklyn and Queens Dominate the Final-Sale Pool
</B> <BR>

Brooklyn: 38.1% <BR>
Queens: 29.1% <BR>
Bronx: 13.6% <BR>
Manhattan: only 10.9% <BR>

In [22]:
# Building class composition

building_class_summary = (
    master_property["building_class"]
    .value_counts(dropna=False)
    .head(25)
    .rename_axis("building_class")
    .reset_index(name="properties")
)

building_class_summary["share"] = (
    building_class_summary["properties"] /
    building_class_summary["properties"].sum()
)

display(building_class_summary)

,building_class,properties,share
0,C0,796,0.099562
1,R4,633,0.079174
2,B2,630,0.078799
3,B1,619,0.077423
4,V0,615,0.076923
5,B3,498,0.062289
6,RG,420,0.052533
7,A1,398,0.049781
8,C3,352,0.044028
9,C2,327,0.040901


In [23]:
# PLUTO land use composition

landuse_summary = (
    master_property["landuse"]
    .value_counts(dropna=False)
    .rename_axis("landuse")
    .reset_index(name="properties")
)

landuse_summary["share"] = (
    landuse_summary["properties"] /
    landuse_summary["properties"].sum()
)

display(landuse_summary)

,landuse,properties,share
0,1.0,2926,0.299703
1,NaN,2033,0.208235
2,2.0,1876,0.192154
3,4.0,906,0.092799
4,11.0,853,0.087371
5,5.0,393,0.040254
6,10.0,264,0.027041
7,6.0,183,0.018744
8,7.0,130,0.013316
9,8.0,116,0.011882


In [24]:
# Residential unit distribution

master_property["unitsres"] = pd.to_numeric(
    master_property["unitsres"],
    errors="coerce"
)

master_property["unitstotal"] = pd.to_numeric(
    master_property["unitstotal"],
    errors="coerce"
)

display(
    master_property[[
        "unitsres",
        "unitstotal"
    ]].describe()
)

,unitsres,unitstotal
count,7776.000000,7776.000000
mean,3.002443,3.369727
std,9.030559,9.144739
min,0.000000,0.000000
25%,0.000000,1.000000
50%,2.000000,2.000000
75%,3.000000,3.000000
max,355.000000,355.000000


In [25]:
# Small homes vs larger buildings

master_property["small_property_flag"] = np.where(
    master_property["unitsres"].fillna(0) <= 3,
    1,
    0
)

small_property_summary = (
    master_property["small_property_flag"]
    .value_counts(dropna=False)
    .rename_axis("small_property_flag")
    .reset_index(name="properties")
)

small_property_summary["share"] = (
    small_property_summary["properties"] /
    small_property_summary["properties"].sum()
)

display(small_property_summary)

,small_property_flag,properties,share
0,1,8283,0.848407
1,0,1480,0.151593


<b> Small Residential Buildings Dominate </b> <br>

C0/C1/C2/C3/C7 → walk-up apartments <br>
B1/B2/B3 → 2-family homes <br>
A1/A2/A5 → 1-family homes <br>
R4 → condo-related <br>
S2 → mixed-use residential  <br>

Most final-sale properties are very small properties (84%)

In [26]:
# Property value distribution

value_cols = [
    "fullval",
    "avtot",
    "avland",
    "extot"
]

for col in value_cols:
    master_property[col] = pd.to_numeric(
        master_property[col],
        errors="coerce"
    )

display(
    master_property[value_cols].describe()
)

,fullval,avtot,avland,extot
count,9.571000e+03,9.571000e+03,9.571000e+03,9.571000e+03
mean,6.889297e+05,1.278419e+05,3.364386e+04,1.200426e+04
std,1.382257e+06,5.586483e+05,1.210361e+05,3.614989e+05
min,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00
25%,2.317895e+05,1.846450e+04,4.587500e+03,0.000000e+00
50%,5.090000e+05,3.094100e+04,8.863000e+03,0.000000e+00
75%,7.700000e+05,9.126500e+04,2.205000e+04,0.000000e+00
max,7.532900e+07,3.389805e+07,5.220000e+06,3.389805e+07


In [27]:
# Top building classes by HPD distress

building_hpd_summary = (
    master_property
    .groupby("building_class", as_index=False)
    .agg(
        properties=("bbl", "nunique"),
        avg_open_hpd_violations=("open_hpd_violations", "mean"),
        avg_open_c_violations=("open_c_violations", "mean")
    )
    .sort_values(
        "avg_open_hpd_violations",
        ascending=False
    )
)

display(building_hpd_summary.head(20))

,building_class,properties,avg_open_hpd_violations,avg_open_c_violations
21,C9,4,127.500000,62.250000
22,D1,26,76.923077,33.307692
117,S0,3,73.000000,14.000000
28,D7,18,68.833333,24.222222
17,C4,40,66.700000,15.250000
14,C1,279,64.892473,18.630824
20,C7,89,62.730337,18.179775
19,C6,3,58.666667,3.666667
18,C5,71,53.408451,11.014085
24,D3,11,51.909091,19.090909


Older walk-up classes,older multifamily stock,
certain mixed-use classes,
and elevator/walk-up apartments

associated with extremely high violation counts.